### DISCUSSION TITLE
KubernetesPodOperator with Dynamic Task IDs, can't figure out format for task_instance.xcom_pull()

### DISCUSSION THREAD

**Discussion by hanleybrand** [apache/airflow#31264] [Q&A] [answered]
I'm pretty new to airflow, and I'm working on a medium size data sync pipeline from a vendor to our data warehouse - since there's almost 200 tables that need to be done, I'm planning to implement dag with dynamic task mapping to run the vendor's custom data sync tool, which requires a number of libraries that conflict with airflow, so I'm using the KubernetesPodOperator with an image I built for the tasks. This is working fine, except I can't figure out how to pass the dynamically generated task id to the BashOperator so I can get the xcom result (I have another K8sPodOp dag that is not dynamic and the Xcom task works fine

This is from a toy/test dag that is simply taking a short list of table names and echoing the name to jq >> /airflow/xcom/return.json -- I setup the job image to use the airflow:0 user and have the same directories because I wanted to mount the shared_temp directory and other stuff:

```python
with DAG( 
# ARGUMENTS
) 
as dag:
    tables = ['access_tokens', 'account_users', 'accounts']

    for table in tables:
        task_id = f"table-test-{table}"
        file_out = "/opt/airflow/temp-data/test_out"
        json_result = f"{file_out}/{table}.json"
        jq = " jq  --raw-input --slurp '. | {schema: .  }'"
        xcom_push = "/airflow/xcom_push/return.json"
        bash_cmd = f"echo -n '{table}' | {jq} | "

        pod_task = KubernetesPodOperator(
            name=f"dap-tabletest-{table}-runner",
            task_id=task_id, 
            image="harbor.k8s.temple.edu/academic-apps/cd2-dap:latest",
            cmds=["bash", "-cx", f"echo -n '{table}' | {jq} > {json_result}; cat {json_result} >> {xcom_push}"],
            do_xcom_push=True,
            # ...ADDITIONAL_ARGUMENTS...
        )
 
        xcom_pull_arg = f"task_instance.xcom_pull('{task_id}')[0]"
        xcom_bash_comm = f"echo \"{{{ {xcom_pull_arg.__str__()} }}}\""

        dap_schema_xcom_result = BashOperator(
            bash_command=xcom_bash_comm[0],
            task_id=f"dap_schema_{table}_xcom_result",
        )
      # returns 
```
I've tried a few different things to pass the task id to `task_instance.xcom_pull()`, but no luck so far - this is the closest I've come:

```python
    xcom_pull_arg = f"task_instance.xcom_pull('{task_id}')[0]"
    # "task_instance.xcom_pull('table-test-access_tokens')[0]"
    
     bash_command=f"echo \"{{ {xcom_pull_arg} }}\"",
     # Xcom return_value:  { task_instance.xcom_pull('table-test-access_tokens') }
```

As far as I can tell, I'm just echoing the text of my command instead of it being executed =\

__________

Note - the xcom step is based on the xcom example in [the KubernetesPodoperator docs](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/operators.html#how-does-xcom-work)

```python
    pod_task_xcom_result = BashOperator(
        bash_command="echo \"{{ task_instance.xcom_pull('write-xcom')[0] }}\"",
        task_id="pod_task_xcom_result",
    )
```

**Comment by potiuk** [comment, marked as answer]
This should do:

```
     bash_command=f"echo \"{{{{ {xcom_pull_arg} }}}}\"",
```

Since you have f-string, the  first `{` in your string is interpreted as a beginning of f-string expression, so what you really get is to print  a set of values returned by another expression `{xcom_pull_arg}` (which evaluest to task_instance.xcom_pull...)

If you want to have `{` as literal in f-string you need to repeat it so `{{{{` in f-string is converted to `{{`  and this is what you want becasue this is what should be passed to a JINJA string.

**Reply by hanleybrand** [reply]
Thanks @potiuk, you're a champion!  I got it to work

After it worked I realized that I had misunderstood the example in the docs -- I had read the ["how does Xcom work?"](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/operators.html#how-does-xcom-work) section as saying that the second step was necessary to extract the xcom return_value, but I suppose it was really just meant to show that it's necessary to use `do_xcom_push=True` and to echo the task output to the return.json , and the second task is meant as a throw-away "then you could theoretically do something like this" example.

Thanks again for your help!

**Reply by potiuk** [reply]
Glad I could help ! 

BTW. If you have an idea to update the docs with more examples and better explanation for others - feel absolutely free to add it. It is super-easy actually. Just click "Suggest a change on this page" and you will get a PR opened where you will be able to edit the docs .rst sources and add whatever you think might explain things better :).
